arangod/RocksDBEngine/RocksDBCollectionMeta.cpp

```diff
@@ -0,0 +1,419 @@
+        return res.reset(rocksutils::convertStatus(s));
+      }
+    }
+  }
+  
+  return res;
+}
+
+/// @brief deserialize collection metadata, only called on startup
+Result RocksDBCollectionMeta::deserializeMeta(rocksdb::DB* db, LogicalCollection& coll) {
+  RocksDBCollection* rcoll = static_cast<RocksDBCollection*>(coll.getPhysical());
+  
+  // Step 1. load the counter
+  auto cf = RocksDBColumnFamily::definitions();
+  rocksdb::ReadOptions ro;
+  ro.fill_cache = false;
+  
+  RocksDBKey key;
+  key.constructCounterValue(rcoll->objectId());
+  
+  rocksdb::PinnableSlice value;
+  rocksdb::Status s = db->Get(ro, cf, key.string(), &value);
+  if (s.ok()) {
+    VPackSlice countSlice = RocksDBValue::data(value);
+    _count = RocksDBCollectionMeta::DocCount(countSlice);
+  } else if (!s.IsNotFound()) {
+    return rocksutils::convertStatus(s);
+  }
+  
+  // Step 2. load the key generator
+  KeyGenerator* keyGen = coll.keyGenerator();
+  if (keyGen->hasDynamicState()) {
+    // only a key generator with dynamic data needs to be recovered
+    key.constructKeyGeneratorValue(rcoll->objectId());
+    s = db->Get(ro, cf, key.string(), &value);
+    if (s.ok()) {
+      
+      VPackSlice keyGenProps = RocksDBValue::data(value);
+      TRI_ASSERT(keyGenProps.isObject());
+      // simon: wtf who decided this is a good deserialization routine ?!
+      VPackSlice val = keyGenProps.get(StaticStrings::LastValue);
+      if (val.isString()) {
+        VPackValueLength size;
+        const char* data = val.getString(size);
+        keyGen->track(data, size);
+      } else if (val.isInteger()) {
+        uint64_t lastValue = val.getUInt();
+        std::string str = std::to_string(lastValue);
+        keyGen->track(str.data(), str.size());
+      }
+      
+    } else if (!s.IsNotFound()) {
+      return rocksutils::convertStatus(s);
+    }
+  }
+
+  // Step 3. load the index estimates
+  auto indexes = coll.getIndexes();
+  for (std::shared_ptr<arangodb::Index>& index : indexes) {
+    RocksDBIndex* idx = static_cast<RocksDBIndex*>(index.get());
+    if (idx->estimator() == nullptr) {
+      LOG_DEVEL << "skipping estimator " << idx->typeName();
+      continue;
+    }
+    
+    key.constructIndexEstimateValue(idx->objectId());
+    s = db->Get(ro, cf, key.string(), &value);
+    if (!s.ok() && !s.IsNotFound()) {
+      return rocksutils::convertStatus(s);
+    } else if (s.IsNotFound()) { // expected with nosync recovery tests
+      LOG_TOPIC(WARN, Logger::ROCKSDB) << "recalculating index estimate for index "
+      << "type '" << idx->typeName() << "' with id '" << idx->id() << "'";
+      idx->recalculateEstimates();
+      continue;
+    }
+    
+    StringRef estimateInput(value.data() + sizeof(uint64_t),
+                            value.size() - sizeof(uint64_t));
+    
+    uint64_t committedSeq = rocksutils::uint64FromPersistent(value.data());
+    if (RocksDBCuckooIndexEstimator<uint64_t>::isFormatSupported(estimateInput)) {
```

Comment:
what do we do if the format is not supported? is this supposed to happen at all?